src/Features/CloudShadows.cpp

```diff
@@ -0,0 +1,281 @@
+			ImGui::Text("This setting affects distortion of clouds near horizon.");
+		}
+		ImGui::TreePop();
+	}
+}
+
+void CloudShadows::CheckResourcesSide(int side)
+{
+	static FrameChecker frame_checker[6];
+	if (!frame_checker[side].isNewFrame())
+		return;
+
+	auto context = RE::BSGraphics::Renderer::GetSingleton()->GetRuntimeData().context;
+
+	float black[4] = { 0, 0, 0, 0 };
+	context->ClearRenderTargetView(cubemapCloudOccRTVs[side], black);
+}
+
+void CloudShadows::ModifySky(const RE::BSShader*, const uint32_t descriptor)
+{
+	auto shadowState = RE::BSGraphics::RendererShadowState::GetSingleton();
+	auto cubeMapRenderTarget = !REL::Module::IsVR() ? shadowState->GetRuntimeData().cubeMapRenderTarget : shadowState->GetVRRuntimeData().cubeMapRenderTarget;
+	if (cubeMapRenderTarget != RE::RENDER_TARGETS_CUBEMAP::kREFLECTIONS)
+		return;
+
+	enum class SkyShaderTechniques
+	{
+		SunOcclude = 0,
+		SunGlare = 1,
+		MoonAndStarsMask = 2,
+		Stars = 3,
+		Clouds = 4,
+		CloudsLerp = 5,
+		CloudsFade = 6,
+		Texture = 7,
+		Sky = 8,
+	};
+
+	auto tech_enum = static_cast<SkyShaderTechniques>(descriptor);
+	if (tech_enum == SkyShaderTechniques::Clouds || tech_enum == SkyShaderTechniques::CloudsLerp || tech_enum == SkyShaderTechniques::CloudsFade) {
+		auto context = RE::BSGraphics::Renderer::GetSingleton()->GetRuntimeData().context;
+		auto renderer = RE::BSGraphics::Renderer::GetSingleton();
+		auto device = renderer->GetRuntimeData().forwarder;
+
+		{
+			ID3D11ShaderResourceView* srv = nullptr;
+			context->PSSetShaderResources(40, 1, &srv);
+		}
+
+		auto reflections = renderer->GetRendererData().cubemapRenderTargets[RE::RENDER_TARGET_CUBEMAP::kREFLECTIONS];
+
+		// render targets
+		ID3D11RenderTargetView* rtvs[4];
+		ID3D11DepthStencilView* depthStencil;
+		context->OMGetRenderTargets(3, rtvs, &depthStencil);
+
+		int side = -1;
+		for (int i = 0; i < 6; ++i)
+			if (rtvs[0] == reflections.cubeSideRTV[i]) {
+				side = i;
+				break;
+			}
+		if (side == -1)
+			return;
+
+		CheckResourcesSide(side);
+
+		rtvs[3] = cubemapCloudOccRTVs[side];
+		context->OMSetRenderTargets(4, rtvs, depthStencil);
+
+		// blend states
+		ID3D11BlendState* blendState;
+		FLOAT blendFactor[4];
+		UINT sampleMask;
+
+		context->OMGetBlendState(&blendState, blendFactor, &sampleMask);
+
+		if (!mappedBlendStates.contains(blendState)) {
+			if (!modifiedBlendStates.contains(blendState)) {
+				D3D11_BLEND_DESC blendDesc;
+				blendState->GetDesc(&blendDesc);
+
+				blendDesc.RenderTarget[3] = blendDesc.RenderTarget[0];
+
+				ID3D11BlendState* modifiedBlendState;
+				DX::ThrowIfFailed(device->CreateBlendState(&blendDesc, &modifiedBlendState));
+
+				mappedBlendStates.insert(modifiedBlendState);
+				modifiedBlendStates.insert({ blendState, modifiedBlendState });
+			}
+			blendState = modifiedBlendStates[blendState];
+			context->OMSetBlendState(blendState, blendFactor, sampleMask);
+		}
+
+		// auto state = RE::BSGraphics::RendererShadowState::GetSingleton();
```

Comment:
Dead code should be removed. It's ok to keep it but normally explaining what it does is helpful.

Reply:
removed